Allow ability to delete multiple instances.

rnatarajan opened the pull request:
Create an API to delete multiple Instances

I am creating this PR to initiate conversation about adding this feature to Cloudbreak.
This feature is needed for teams using YARN Nodel Labels and attempting to delete multiple InstanceIds.

**Motivation:**
[Yarn Node Labels](https://hadoop.apache.org/docs/r2.7.3/hadoop-yarn/hadoop-yarn-site/NodeLabel.html) are added since Hadoop 2.7.3
Once the cluster is scaled up for a specific Yarn Node Label, use case arises to delete specific instances rather than use Cloudbreak Downscale or resize.
Cloudbreak Downscale would delete arbitrary Instances where as Yarn Node Labeling creates the use case of deleting specific instance ids.

**Solution:**
Create an API to allow Cloudbreak to delete multiple Instance Ids.
As a hack, i have removed the check that prevents deletion of Healthy instances(I can fix this).
But if Hortonworks is ready to allow this feature(delete multiple instanceIds), i can add a boolean flag(default=false) to the API and the instances can only be deleted if the boolean flag is enabled.

**Issues with current code:**
1. Deleting an instance takes about a minute to two and a team attempting to delete 100 instances using delete instance API would take about 1.6 - 3.33 hours.
2. Alternate solutions provided by Hortonworks members in cloudbreak forum to delete the instances from EC2 and call sync in cloudbreak turned futile. - Cloudbreak DB and Ambari DB were not updated with correct status for the instances. And when Cloudbreak Sync is used AWS Autoscaling was not updated with correct number causing upscale/downscale to fail.

**TO DO:**
1. Remove the Hack to delete Healthy instance and add overriding boolean flag, which will be disabled by default.
2. Write test cases covering all scenarios for deleteInstanceIds.

keyki commented on core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/service/ReactorFlowManager.java:
```diff
@@ -112,6 +113,13 @@ public void triggerStackRemoveInstance(Long stackId, String hostGroup, Long priv
 
+    public void triggerStackRemoveInstances(Long stackId, String hostGroup, Set<String> hostnames) {
+        String selector = FlowChainTriggers.FULL_DOWNSCALE_TRIGGER_EVENT;
+        ClusterAndStackDownscaleTriggerEvent event = new ClusterAndStackDownscaleTriggerEvent(selector, stackId, hostGroup, hostnames,
```
There is no constructor for setting the hostnames, only the CB private id is supported for now. This causes a compile error.

keyki commented on core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java:
```diff
@@ -434,6 +437,30 @@ public void removeInstance(IdentityUser user, Long stackId, String instanceId) {
     }
 
+    public void removeInstances(IdentityUser user, Long stackId, Set<String> instanceIds) {
+        Stack stack = get(stackId);
```
The get method checks for permissions (authorizationService.hasReadPermission(stack);), but this method should be extended with hasWritePermission instead of manually checking the owner and public in account field below.

rnatarajan commented:
@keyki Thanks a lot for the review. I will address the review comments and will fix the TODOs that I have listed in the PR.

Regarding your questions,

**do you use the API directly to send the requests or cli, maybe a custom tool?**
- We use Cloudbreak API's. Our application uses cloud-api as a dependency and the applicaiton calls cloudbreak to perform autoscaling.
- Also providing this feature in cloudbreak with a warning will be useful.

do you have datanodes on the labeled yarn nodes? this is important, because normally CB tries to find those instances which have the least amount of data within the host group you try to downscale and also initiates a decommission in NN. if you specify the instanceids this won't happen and no data will be moved
- No we do not have datanodes on labeled YARN Nodes.
- The worker nodes are labeled and the worker nodes have HDFS Client, Hive Client, Mapreduce Client, Node Manager Client, Metics Monitor, Oozie Client, Pig Client, Slider Client, Spark Client, Tez Client, YARN Client.
- Worker nodes perform only the mapreduce/spark/tez container work and there are DataNodes dedicated in the cluster for HDFS.

keyki commented:
@rnatarajan We discussed your feature and we would like to merge it, once the issues are fixed. One question though. You are using 2.4 which is a released version and we rarely do feature development on released versions, most of the time these are bugfixes or backports. Are you open to upgrade to a newer version or you definitely need this feature on 2.4?

rnatarajan commented:
Yes we are open to upgrade to the new version. 
we went from 1.16 to 2.4 as soon as you released 2.4. And we always stay on the most recent release.
I will create a PR to the tip of master with this feature addressing the review comments.

Thanks a lot for the quick response and review.

keyki commented:
Closing this prio the other PR
